src/gluonnlp/model/bert.py

```diff
@@ -167,40 +96,158 @@ class BERTEncoder(BaseTransformerEncoder):
-                                          prefix=prefix, params=params,
-                                          # extra configurations for BERT
-                                          positional_weight='learned',
-                                          use_bert_encoder=True,
-                                          use_layer_norm_before_dropout=False,
-                                          scale_embed=False,
-                                          activation=activation,
-                                          layer_norm_eps=layer_norm_eps)
-
-
-class BERTEncoderCell(BaseTransformerEncoderCell):
-    """Structure of the Transformer Encoder Cell for BERT.
-
-    Different from the original encoder cell for transformer,
-    `BERTEncoderCell` adds bias terms for attention and the projection
-    on attention output. It also uses `BERTPositionwiseFFN` and
-    `BERTLayerNorm`.
+    def __init__(self, *, attention_cell='multi_head', num_layers=2, units=512, hidden_size=2048,
+                 max_length=50, num_heads=4, scaled=True, dropout=0.0, use_residual=True,
+                 output_attention=False, output_all_encodings=False, weight_initializer=None,
+                 bias_initializer='zeros', prefix=None, params=None, activation='gelu',
+                 layer_norm_eps=1e-12):
+        super().__init__(prefix=prefix, params=params)
+        assert units % num_heads == 0,\
+            'In BERTEncoder, The units should be divided exactly ' \
+            'by the number of heads. Received units={}, num_heads={}' \
+            .format(units, num_heads)
+        self._max_length = max_length
+        self._units = units
+        self._output_attention = output_attention
+        self._output_all_encodings = output_all_encodings
+        self._dropout = dropout
+
+        with self.name_scope():
+            if dropout:
+                self.dropout_layer = nn.Dropout(rate=dropout)
+            self.layer_norm = nn.LayerNorm(in_channels=units, epsilon=1e-12)
+            self.position_weight = self.params.get('position_weight', shape=(max_length, units),
+                                                   init=weight_initializer)
+            self.transformer_cells = nn.HybridSequential()
+            for i in range(num_layers):
+                cell = BERTEncoderCell(
+                    units=units, hidden_size=hidden_size, num_heads=num_heads,
+                    attention_cell=attention_cell, weight_initializer=weight_initializer,
+                    bias_initializer=bias_initializer, dropout=dropout, use_residual=use_residual,
+                    scaled=scaled, output_attention=output_attention, prefix='transformer%d_' % i,
+                    activation=activation, layer_norm_eps=layer_norm_eps)
+                self.transformer_cells.add(cell)
+
+    def __call__(self, inputs, states=None, valid_length=None): #pylint: disable=arguments-differ
```

Comment:
My comment is not related to refactoring: is there any issue using the recently added None support in gluon, and removing `__call__`? Just curious

Reply:
Gluon doesn't support keyword arguments. So removing call will break API. We should add keyword argument support to Gluon..